pkg/vault/test/foundry/VaultExplorer.t.sol

```diff
@@ -0,0 +1,750 @@
+
+        for (uint256 i = 0; i < decimalScalingFactors.length; ++i) {
+            assertEq(
+                decimalScalingFactors[i],
+                10 ** (18 + tokenDecimalDiffs[i]),
+                string.concat("decimalScalingFactors of token", Strings.toString(i), "should match tokenDecimalDiffs")
+            );
+
+            assertEq(
+                tokenRates[i],
+                rateProviders[i].getRate(),
+                string.concat("tokenRate of token", Strings.toString(i), "does not match mock providers.")
+            );
+        }
+    }
+
+    function testGetPoolData() public view {
+        PoolData memory poolData = explorer.getPoolData(pool);
+        IERC20[] memory tokens = vault.getPoolTokens(pool);
+
+        assertTrue(poolData.tokenInfo[daiIdx].paysYieldFees, "DAI doesn't pay yield fees");
+        assertFalse(poolData.tokenInfo[usdcIdx].paysYieldFees, "USDC pays yield fees");
+
+        assertTrue(poolData.poolConfigBits.isPoolRegistered(), "Pool not registered");
+        assertTrue(poolData.poolConfigBits.isPoolInitialized(), "Pool not registered");
+
+        assertEq(poolData.balancesRaw[daiIdx], DAI_RAW_BALANCE, "DAI raw balance wrong");
+        assertEq(poolData.balancesRaw[usdcIdx], USDC_RAW_BALANCE, "USDC raw balance wrong");
+
+        uint256 daiLiveBalance = poolData.balancesRaw[daiIdx].toScaled18ApplyRateRoundDown(
+            poolData.decimalScalingFactors[daiIdx],
+            poolData.tokenRates[daiIdx]
+        );
+        uint256 usdcLiveBalance = poolData.balancesRaw[usdcIdx].toScaled18ApplyRateRoundDown(
+            poolData.decimalScalingFactors[usdcIdx],
+            poolData.tokenRates[usdcIdx]
+        );
+
+        assertEq(poolData.balancesLiveScaled18[daiIdx], daiLiveBalance, "DAI live balance wrong");
+        assertEq(poolData.balancesLiveScaled18[usdcIdx], usdcLiveBalance, "USDC live balance wrong");
+
+        for (uint256 i = 0; i < tokens.length; ++i) {
+            assertEq(
+                poolData.decimalScalingFactors[i],
+                10 ** (18 + tokenDecimalDiffs[i]),
+                string.concat("decimalScalingFactors of token ", Strings.toString(i), " should match tokenDecimalDiffs")
+            );
+
+            assertEq(
+                poolData.tokenRates[i],
+                rateProviders[i].getRate(),
+                string.concat("tokenRate of token ", Strings.toString(i), " does not match mock providers.")
+            );
+
+            assertEq(
+                address(poolData.tokens[i]),
+                address(tokens[i]),
+                string.concat("Address of token ", Strings.toString(i), " does not match.")
+            );
+
+            assertEq(
+                uint8(poolData.tokenInfo[i].tokenType),
+                uint8(TokenType.WITH_RATE),
+                string.concat("Token type of token ", Strings.toString(i), " does not match")
+            );
+        }
```

Comment:
I think you could also get pool data from vault, then from explorer, and compare the hashes right?

Reply:
I guess, but this would tell you exactly what was wrong, and it's kind of self-documenting.